lang/LangSource/PyrSlot.h

```diff
@@ -27,16 +27,398 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
+        return { PrivateTag(), symTag | static_cast<uint64_t>(reinterpret_cast<uintptr_t>(o)) };
+    }
+    [[nodiscard]] static PyrSlot make(int32_t i) {
+        return { PrivateTag(), intTag | static_cast<uint64_t>(bit_cast<uint32_t>(i)) };
+    }
+    [[nodiscard]] static PyrSlot make(void* o) {
+        return { PrivateTag(), ptrTag | static_cast<uint64_t>(reinterpret_cast<uintptr_t>(o)) };
+    }
+    [[nodiscard]] constexpr static PyrSlot make(PyrNil) noexcept { return {}; }
+    [[nodiscard]] constexpr static PyrSlot make(bool b) noexcept {
+        return { PrivateTag(), b ? trueValue : falseValue };
+    }
+
+    [[nodiscard]] bool isDouble() const noexcept { return !isBoxed(); }
+    [[nodiscard]] bool isChar() const noexcept { return tagChecker<charTag>(); }
+    [[nodiscard]] bool isInt() const noexcept { return tagChecker<intTag>(); }
+    [[nodiscard]] bool isPtr() const noexcept { return tagChecker<ptrTag>(); }
+    [[nodiscard]] bool isObjectHdr() const noexcept { return tagChecker<objHdrTag>(); }
+    [[nodiscard]] bool isSymbol() const noexcept { return tagChecker<symTag>(); }
+    [[nodiscard]] bool isNil() const noexcept { return u_raw == nilValue; }
+    [[nodiscard]] bool isTrue() const noexcept { return u_raw == trueValue; }
+    [[nodiscard]] bool isFalse() const noexcept { return u_raw == falseValue; }
+
+    [[nodiscard]] double getDouble() const noexcept {
+        assert(isDouble());
+        return u_double;
+    }
+    [[nodiscard]] char getChar() const noexcept {
+        assert(isChar());
+        return u_char.value;
+    }
+    [[nodiscard]] int32_t getInt() const noexcept {
+        assert(isInt());
+        return u_int.value;
+    }
+    [[nodiscard]] void* getPtr() const noexcept {
+        assert(isPtr());
+        return u_ptr.getPtr();
+    }
+    [[nodiscard]] struct PyrObjectHdr* getObjectHdr() const noexcept {
+        assert(isObjectHdr());
+        return u_objectHeader.getPtr();
+    }
+    [[nodiscard]] struct PyrSymbol* getSymbol() const noexcept {
+        assert(isSymbol());
+        return u_symbol.getPtr();
+    }
+    template <typename T> [[nodiscard]] T* get() const noexcept {
+        // types are incomplete here so can't check...
+        // static_assert(std::is_base_of<PyrObjectHdr, T>::value, "Type must derive from PyrObjectHeader");
+        if (isObjectHdr())
+            return reinterpret_cast<T*>(u_objectHeader.getPtr());
+        assert(isNil() || (isInt() && getInt() == 0));
+        return nullptr;
+    }
+};
+
+static_assert(sizeof(PyrSlot) == sizeof(double));
+
+
+// ORDER IS IMPORTANT AND ASSUMED THROUGHOUT --- DO NOT CHANGE!!!
+enum {
+    tagNotInitialized, // uninitialized slots have a tag of 0
+    tagObj,
+    tagInt,
+    tagSym,
+    tagChar,
+    tagNil,
+    tagFalse,
+    tagTrue,
+    tagPtr,
+    tagFloat,
+    tagUnused,
+};
+
+inline int GetTag(const PyrSlot* slot) noexcept {
+    if (slot->isChar())
```

Comment:
I wonder if we should move nil/bool/int cases to the top here, since they are more likely to be performance critical and common? Actually, an empirical answer to "which of these is called the most" would be nice to have.

Reply:
I'll take some measurements! (Probably be next week)

Reply:
I've changed this now to better reflect the structure of the tags, using `if`s for the values (nil, true, and false) and a switch on the remaining tags.